server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Scanner.java

```diff
@@ -38,100 +40,126 @@
   private ScanDataSource isolatedDataSource;
   private boolean sawException = false;
   private boolean scanClosed = false;
+  protected Semaphore scannerSemaphore;
 
   Scanner(Tablet tablet, Range range, ScanOptions options) {
     this.tablet = tablet;
     this.range = range;
     this.options = options;
+    scannerSemaphore = new Semaphore(1, true);
   }
 
-  public synchronized ScanBatch read() throws IOException, TabletClosedException {
-
-    if (sawException)
-      throw new IllegalStateException("Tried to use scanner after exception occurred.");
+  public ScanBatch read() throws IOException, TabletClosedException {
 
     if (scanClosed)
```

Comment:
Below is from the semaphore javadoc.   Reading that is sounds like there is no guarantee of seeing a change to `scanClosed` until after `acquire()`.

```
Memory consistency effects: Actions in a thread prior to calling a "release" method such as release() happen-before actions following a successful "acquire" method such as acquire() in another thread.
```